include/buffer.h

```diff
@@ -53,53 +61,24 @@ class buffer {
-		using rmfn_traits = allscale::utils::lambda_traits<Functor>;
-		static_assert(rmfn_traits::result_type::dims == Dims, "The returned subrange doesn't match buffer dimensions.");
-
-		if(detail::is_prepass_handler(cgh)) {
-			auto& prepass_cgh = dynamic_cast<detail::prepass_handler&>(cgh);
-			prepass_cgh.add_requirement(id, std::make_unique<detail::range_mapper<rmfn_traits::arg1_type::dims, Dims>>(rmfn, Mode, get_range()));
-			if constexpr(Target == cl::sycl::access::target::host_buffer) {
-				return detail::make_host_accessor<DataT, Dims, Mode>();
-			} else {
-				return detail::make_device_accessor<DataT, Dims, Mode>(*faux_buf);
-			}
-		}
 
-		// It's difficult to figure out which stored range mapper corresponds to this get_access call, which is why we just call the raw mapper manually.
-		// This also means that we have to clamp the subrange ourselves here, which is not ideal from an encapsulation standpoint.
-		if constexpr(Target == cl::sycl::access::target::host_buffer) {
-			if(detail::get_handler_execution_target(cgh) != detail::execution_target::HOST) {
-				throw std::runtime_error("Calling buffer::get_access with sycl::access::target::host_buffer is only allowed in host tasks.");
-			}
-			auto& live_cgh = dynamic_cast<detail::live_pass_host_handler&>(cgh);
-			const auto sr = detail::clamp_subrange_to_buffer_size(live_cgh.apply_range_mapper<Dims>(rmfn, get_range()), get_range());
-			auto access_info = detail::runtime::get_instance().get_buffer_manager().get_host_buffer<DataT, Dims>(
-			    id, Mode, detail::range_cast<3>(sr.range), detail::id_cast<3>(sr.offset));
-			return detail::make_host_accessor<DataT, Dims, Mode>(sr, access_info.buffer, access_info.offset, range);
-		} else {
-			if(detail::get_handler_execution_target(cgh) != detail::execution_target::DEVICE) {
-				throw std::runtime_error(
-				    "Calling buffer::get_access on device buffers is only allowed in compute tasks. "
-				    "If you want to access this buffer from within a host task, please specialize the call using sycl::access::target::host_buffer.");
-			}
-			auto& live_cgh = dynamic_cast<detail::live_pass_device_handler&>(cgh);
-			const auto sr = detail::clamp_subrange_to_buffer_size(live_cgh.apply_range_mapper<Dims>(rmfn, get_range()), get_range());
-			auto access_info = detail::runtime::get_instance().get_buffer_manager().get_device_buffer<DataT, Dims>(
-			    id, Mode, detail::range_cast<3>(sr.range), detail::id_cast<3>(sr.offset));
-			return detail::make_device_accessor<DataT, Dims, Mode>(live_cgh.get_eventual_sycl_cgh(), sr, access_info.buffer, access_info.offset);
-		}
+	template <cl::sycl::access_mode Mode, cl::sycl::target Target, typename Functor>
+	auto get_access(handler& cgh, Functor rmfn) const {
+		return accessor<DataT, Dims, Mode, Target>(*this, cgh, rmfn);
+	}
+
+#if WORKAROUND_COMPUTECPP
+	template <cl::sycl::access_mode Mode, cl::sycl::access::target Trgt, typename Functor>
+	auto get_access(handler& cgh, Functor rmfn) const {
```

Comment:
Same here.

Reply:
done in new commit.

Reply:
We're keeping this deduced return type because we would otherwise implement some additional translation for `cl::sycl::access::target`, and this workaround is temporary anyway.